src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -710,6 +733,130 @@ - (BOOL)_attributeDataValue:(MTRDeviceDataValueDictionary)observed satisfiesValu
+
+    std::lock_guard lock(_lock);
+    auto * attributeWaiter = [[MTRAttributeValueWaiter alloc] initWithDevice:self values:values queue:queue completion:completion];
+
+    for (MTRDeviceResponseValueDictionary currentValue in currentValues) {
+        // Pretend as if this got reported, for purposes of the attribute
+        // waiter.
+        [attributeWaiter _attributeValue:currentValue[MTRDataKey] reportedForPath:currentValue[MTRAttributePathKey] byDevice:self];
+    }
+
+    if (attributeWaiter.allValuesSatisfied) {
+        MTR_LOG("%@ waitForAttributeValues no need to wait, values already match: %@", self, values);
+        [attributeWaiter _notifyWithError:nil];
+        return attributeWaiter;
+    }
+
+    // Otherwise, wait for the values to arrive or our timeout.
+    if (!self.attributeValueWaiters) {
+        self.attributeValueWaiters = [NSHashTable weakObjectsHashTable];
+    }
+    [self.attributeValueWaiters addObject:attributeWaiter];
+
+    MTR_LOG("%@ waitForAttributeValues will wait up to %f seconds for %@", self, timeout, values);
+    dispatch_source_t timerSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, self.queue);
+    attributeWaiter.expirationTimer = timerSource;
+
+    // Set a timer to go off after timeout, and not repeat.
+    dispatch_source_set_timer(timerSource, dispatch_time(DISPATCH_TIME_NOW, static_cast<uint64_t>(timeout * static_cast<double>(NSEC_PER_SEC))), DISPATCH_TIME_FOREVER,
+        // Allow .5 seconds of leeway; should be plenty, in practice.
+        static_cast<uint64_t>(0.5 * static_cast<double>(NSEC_PER_SEC)));
+
+    mtr_weakify(attributeWaiter);
+    mtr_weakify(self);
+    mtr_weakify(timerSource);
+    dispatch_source_set_event_handler(timerSource, ^{
+        mtr_strongify(self);
+        mtr_strongify(attributeWaiter);
+        mtr_strongify(timerSource);
+        if (self != nil && attributeWaiter != nil) {
+            [self _attributeWaitTimedOut:attributeWaiter];
+        } else if (timerSource != nil) {
+            // Make sure to cancel the timer ourselves, if we can't call _attributeWaitTimedOut.
+            dispatch_source_cancel(timerSource);
+        }
+    });
+
+    dispatch_resume(timerSource);
+    return attributeWaiter;
+}
+
+- (void)_attributeValue:(MTRDeviceDataValueDictionary)value reportedForPath:(MTRAttributePath *)path
+{
+    os_unfair_lock_assert_owner(&_lock);
+
+    // Check whether anyone was waiting for this attribute.
+    NSHashTable * satisfiedWaiters = [NSHashTable weakObjectsHashTable];
+    for (MTRAttributeValueWaiter * attributeValueWaiter in self.attributeValueWaiters) {
+        if ([attributeValueWaiter _attributeValue:value reportedForPath:path byDevice:self] && attributeValueWaiter.allValuesSatisfied) {
+            [satisfiedWaiters addObject:attributeValueWaiter];
+        }
+    }
+
+    [self.attributeValueWaiters minusHashTable:satisfiedWaiters];
```

Comment:
```suggestion
    NSMutableArray * satisfiedWaiters;
    for (MTRAttributeValueWaiter * attributeValueWaiter in self.attributeValueWaiters) {
        if ([attributeValueWaiter _attributeValue:value reportedForPath:path byDevice:self] && attributeValueWaiter.allValuesSatisfied) {
            if (!satisfiedWaiters) {
                satisfiedWaiter = [[NSMutableArray alloc] init];
            }
            [satisfiedWaiters addObject:attributeValueWaiter];
        }
    }
```

I think we can avoid allocating the satisfiedWaiters container until needed (the `for in` loop below is a no-op if satisfiedWaiters is nil). Also making it weak seems harder to reason about for no benefit (unless I'm missing something), and `_notify...` already handles the removal from `self.attributeValueWaiters`.

Reply:
It was a weak thing so I could do the minusHashTable.  I'll double-check that the removals happen in some other way, and if so adjust accordingly.